app/services/ProductServiceImpl.java

```diff
@@ -0,0 +1,41 @@
+package services;
+
+import exceptions.ProductNotFoundException;
+import io.sphere.sdk.client.SphereClient;
+import io.sphere.sdk.products.ProductProjection;
+import io.sphere.sdk.products.ProductVariant;
+import io.sphere.sdk.products.queries.ProductProjectionQuery;
+import io.sphere.sdk.queries.PagedQueryResult;
+
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+
+public class ProductServiceImpl extends AbstractShopService implements ProductService {
+
+    private final ProductProjection cachedProduct;
+
+    public ProductServiceImpl(final SphereClient sphereClient) {
+        super(sphereClient);
+        final Optional<ProductProjection> product = Optional.of(loadProduct().orElseThrow(ProductNotFoundException::new));
+        this.cachedProduct = product.get();
+    }
+
+    @Override
+    public Optional<ProductVariant> getVariantFromId(ProductProjection product, int variantId) {
+        return product.getAllVariants().stream().filter(v -> v.getId().equals(variantId)).findFirst();
```

Comment:
you can inline this method with `return Optional.ofNullable(product.getVariant(variantId));` or just `product.getVariant(variantId)` if you call `Optional#get()` anyway.